### DISCUSSION TITLE
No conf t

### DISCUSSION THREAD

**Discussion by andreirapuru** [ktbyers/netmiko#3093] [Q&A] [answered]
Hello there,
is it possible to configure netmiko to not enter "conf t"?
I need to get information (show version, for example) from access-points that are managed by Cisco WLC, and these access-points "doesn't have conf t" mode.
How to do that?

Thanks all.

**Comment by ktbyers** [comment]
@andreirapuru So you are trying to SSH directly into the access-points?

**Reply by andreirapuru** [reply]
Yes, precisely.
And they have user, pass and enable, but no "conf t".

**Comment by andreirapuru** [comment]
Manual access sample:

login as: admin
admin@172.200.164.99's password:
AP001>en
Password:
AP001#show ver
             Restricted Rights Legend

Use, duplication, or disclosure by the Government is subject to
restrictions as set forth in subparagraph (c) of the Commercial
Computer Software - Restricted Rights clause at FAR sec. 52.227-19 and
subparagraph (c) (1) (ii) of the Rights in Technical Data and Computer
Software clause at DFARS sec. 252.227-7013.
<OUTPUT OMITTED>



Logs from Netmiko:
DEBUG:paramiko.transport:starting thread (client mode): 0x458b190
DEBUG:paramiko.transport:Local version/idstring: SSH-2.0-paramiko_3.0.0
DEBUG:paramiko.transport:Remote version/idstring: SSH-2.0-OpenSSH_7.8
INFO:paramiko.transport:Connected (version 2.0, client OpenSSH_7.8)
DEBUG:paramiko.transport:=== Key exchange possibilities ===
DEBUG:paramiko.transport:kex algos: ecdh-sha2-nistp256, ecdh-sha2-nistp384, ecdh-sha2-nistp521, diffie-hellman-group18-sha512, diffie-hellman-group14-sha1
DEBUG:paramiko.transport:server key: ssh-rsa, ecdsa-sha2-nistp256
DEBUG:paramiko.transport:client encrypt: aes128-ctr, aes192-ctr, aes256-ctr, aes128-gcm@openssh.com, aes256-gcm@openssh.com
DEBUG:paramiko.transport:server encrypt: aes128-ctr, aes192-ctr, aes256-ctr, aes128-gcm@openssh.com, aes256-gcm@openssh.com
DEBUG:paramiko.transport:client mac: hmac-sha2-256-etm@openssh.com, hmac-sha2-512-etm@openssh.com, hmac-sha2-256, hmac-sha2-512
DEBUG:paramiko.transport:server mac: hmac-sha2-256-etm@openssh.com, hmac-sha2-512-etm@openssh.com, hmac-sha2-256, hmac-sha2-512
DEBUG:paramiko.transport:client compress: none, zlib@openssh.com
DEBUG:paramiko.transport:server compress: none, zlib@openssh.com
DEBUG:paramiko.transport:client lang: <none>
DEBUG:paramiko.transport:server lang: <none>
DEBUG:paramiko.transport:kex follows: False
DEBUG:paramiko.transport:=== Key exchange agreements ===
DEBUG:paramiko.transport:Kex: ecdh-sha2-nistp256
DEBUG:paramiko.transport:HostKey: ecdsa-sha2-nistp256
DEBUG:paramiko.transport:Cipher: aes128-ctr
DEBUG:paramiko.transport:MAC: hmac-sha2-256
DEBUG:paramiko.transport:Compression: none
DEBUG:paramiko.transport:=== End of kex handshake ===
DEBUG:paramiko.transport:kex engine KexNistp256 specified hash_algo <built-in function openssl_sha256>
DEBUG:paramiko.transport:Switch to new keys ...
DEBUG:paramiko.transport:Got EXT_INFO: {'server-sig-algs': b'ssh-ed25519,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521'}
DEBUG:paramiko.transport:Adding ecdsa-sha2-nistp256 host key for 172.200.164.99: b'1b456fe72e8ba86dd7f08716c0fb63ae'
DEBUG:paramiko.transport:userauth is OK
INFO:paramiko.transport:Authentication (password) successful!
DEBUG:paramiko.transport:[chan 0] Max packet in: 32768 bytes
DEBUG:paramiko.transport:Received global request "hostkeys-00@openssh.com"
DEBUG:paramiko.transport:Rejecting "hostkeys-00@openssh.com" global request from server.
DEBUG:paramiko.transport:[chan 0] Max packet out: 32768 bytes
DEBUG:paramiko.transport:Secsh channel 0 opened.
DEBUG:paramiko.transport:[chan 0] Sesch channel 0 request ok
DEBUG:paramiko.transport:[chan 0] Sesch channel 0 request ok
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:write_channel: b'terminal width 511\n'
DEBUG:netmiko:read_channel: 

DEBUG:netmiko:read_channel: terminal width 511

DEBUG:netmiko:Pattern found: (terminal width 511) 
terminal width 511
DEBUG:netmiko:In disable_paging
DEBUG:netmiko:Command: terminal length 0

DEBUG:netmiko:write_channel: b'terminal length 0\n'
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: terminal length 0

DEBUG:netmiko:Pattern found: (terminal\ length\ 0) 
terminal length 0
DEBUG:netmiko:
terminal length 0
DEBUG:netmiko:Exiting disable_paging
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:Clear buffer detects data in the channel
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 

DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
AP001>
DEBUG:netmiko:Pattern found: (\#|>) 

AP001>
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:[find_prompt()]: prompt is AP001>
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 

AP001>
AP001>
DEBUG:netmiko:Pattern found: ([>#]) 

AP001>
DEBUG:netmiko:write_channel: b'configure terminal\n'
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: configure terminal
               ^
% Invalid input detected at '^' marker.

AP001>
AP001>
DEBUG:netmiko:Pattern found: (configure\ terminal) 
AP001>configure terminal
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:Pattern found: (AP001.*) 
               ^
% Invalid input detected at '^' marker.

AP001>
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:Pattern found: ([>#]) 
AP001>
DEBUG:paramiko.transport:starting thread (client mode): 0x45e10f0
DEBUG:paramiko.transport:Local version/idstring: SSH-2.0-paramiko_3.0.0
DEBUG:paramiko.transport:Remote version/idstring: SSH-2.0-OpenSSH_7.8

**Comment by ktbyers** [comment]
@andreirapuru What operating system do these devices run?

Can you show the `show version` output?

Also do these access-points have output paging (i.e. do they do some form of a `---more---` to page output and how do you disable this)?

**Reply by andreirapuru** [reply]
They run AIR-OS. Yes, they have "paging", and we can disable it with the command "terminal length 0".

AP001#sh ver
             Restricted Rights Legend

Use, duplication, or disclosure by the Government is subject to
restrictions as set forth in subparagraph (c) of the Commercial
Computer Software - Restricted Rights clause at FAR sec. 52.227-19 and
subparagraph (c) (1) (ii) of the Rights in Technical Data and Computer
Software clause at DFARS sec. 252.227-7013.

            Cisco Systems, Inc.
            170 West Tasman Drive
            San Jose, California 95134-1706

This product contains cryptographic features and is subject to United
States and local country laws governing import, export, transfer and
use. Delivery of Cisco cryptographic products does not imply
third-party authority to import, export, distribute or use encryption.
Importers, exporters, distributors and users are responsible for
compliance with U.S. and local country laws. By using this product you
agree to comply with applicable laws and regulations. If you are unable
to comply with U.S. and local laws, return this product immediately.

A summary of U.S. laws governing Cisco cryptographic products may be found at:
http://www.cisco.com/wwl/export/crypto/tool/stqrg.html

If you require further assistance please contact us by sending email to
export@cisco.com.

This product contains some software licensed under the
"GNU General Public License, version 2" provided with
ABSOLUTELY NO WARRANTY under the terms of
"GNU General Public License, version 2", available here:
http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

This product contains some software licensed under the
"GNU Library General Public License, version 2" provided
with ABSOLUTELY NO WARRANTY under the terms of "GNU Library
General Public License, version 2", available here:
http://www.gnu.org/licenses/old-licenses/lgpl-2.0.html

This product contains some software licensed under the
"GNU Lesser General Public License, version 2.1" provided
with ABSOLUTELY NO WARRANTY under the terms of "GNU Lesser
General Public License, version 2.1", available here:
http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html

This product contains some software licensed under the 
"GNU General Public License, version 3" provided with 
ABSOLUTELY NO WARRANTY under the terms of  
"GNU General Public License, Version 3", available here: 
http://www.gnu.org/licenses/gpl.html.

This product contains some software licensed under the
"GNU Affero General Public License, version 3" provided
--More-- 
         
with ABSOLUTELY NO WARRANTY under the terms of
"GNU Affero General Public License, version 3", available here:
http://www.gnu.org/licenses/agpl-3.0.html.

Cisco AP Software, (ap3g3), C2802, RELEASE SOFTWARE
Technical Support: http://www.cisco.com/techsupport
Copyright (c) 1986-2023 by Cisco Systems, Inc.
Compiled Tue Nov 22 06:10:34 PST 2022

ROM: Bootstrap program is U-Boot boot loader
BOOTLDR: U-Boot boot loader Version 2013.01-gcb01933d4 (Nov 10 2020 - 23:41:12)

AP001 uptime is 0 days, 0 hours, 49 minutes
Last reload time   : Tue Dec 27 19:15:35 UTC 2022
Last reload reason : unknown
 
cisco AIR-AP2802I-Z-K9 ARMv7 Processor rev 1 (v7l) with 1028056/516152K bytes of memory.
Processor board ID TSP2223ACUC
AP Running Image     : 8.10.171.0
Primary Boot Image   : 8.10.171.0
Backup Boot Image    : 8.10.171.0
Primary Boot Image Hash: 4b517f58074e63f923dc4f621fe079d1f7718944f3c3392faea7f5b3360d56b230159b2cda1239b82deed879236cd34ad21c464a540868de5ad4
Backup  Boot Image Hash: 932babe8e070ef19eefb12df5d2c8f46c9ccfa4f2e539182fa7029fec445d507a26cc5f20f94d8088aaf27f7d7131c34892d231227824df4e694
2 Gigabit Ethernet interfaces
2 802.11 Radios
Radio Driver version : 9.0.5.5-W8964
Radio FW version : 9.1.8.1
NSS FW version : 2.4.28

Base ethernet MAC Address            : 00:FC:BA:2A:0E:F1
Part Number                          : 73-100821-05
PCA Assembly Number                  : 000-00000-00
PCA Revision Number                  : 
PCB Serial Number                    : FDO22122UD0
Top Assembly Part Number             : 068-100534-01
Top Assembly Serial Number           : TSP2223ACUA
Top Revision Number                  : A0
Product/Model Number                 : AIR-AP2802I-Z-K9

**Reply by ktbyers** [reply]
Does `terminal length 0` work at the lower privilege level i.e. when the prompt is `>`?

**Reply by andreirapuru** [reply]
Yes, it does.

**Comment by ktbyers** [comment, marked as answer]
So the issue is they have this non-standard SSH login:

```
login as: admin
admin@172.200.164.99's password:
AP001>en
```

That login-as requires special code to handle. The WLC does that same `login as`, but then has a different mechanism for disabling output paging (which requires going into config mode--which is why the WLC driver does that configuration behavior).

So we need a solution for how to handle those two sets of behaviors i.e. non-standard SSH login, but standard Cisco output paging disable.

**Reply by andreirapuru** [reply]
Not sure if I understood. From the logs looks like netmiko was able to connect and disable paging. But after that it tried to issue "conf t" and those access-points "don't have" this option.

**Reply by ktbyers** [reply]
My bad...I assumed some invalid things.

Which Netmiko device_type are you using?

Also can you please post your code that is causing the error.

**Reply by andreirapuru** [reply]
No problem!!
I tried cisco_ios and cisco_wlc as device type. The code is basicaly this (always worked for router and switches - traditional IOS): https://github.com/andreirapuru/netmiko_send_commands/blob/main/send_commands.py

**Reply by ktbyers** [reply]
What is in this `commands_to_send.txt` file?

```
output = net_connect.send_config_from_file('commands_to_send.txt')
```

Here you are executing `send_config_from_file` for config changes on `cisco IOS`, Netmiko will always try to enter/exit configuration mode.

If the commands in this file are not configuration commands, then you should use a different Netmiko method here (I can provide more details on which method will be a good fit once I learn some more details on the commands).

**Reply by andreirapuru** [reply]
It could be any command I want to send to devices (usually more than one command, and when I send commands "show" I just use the format "do show ....", so I don't need to exit config mode).
In this specific case it has only "show version" (I need to get show version from 250 access-points).

**Reply by ktbyers** [reply]
The Netmiko process for doing this would either be:

1. Read in the file and then loop over and send them via `send_command()` or
2. Read in the file and then just use send_multiline()

**Reply by ktbyers** [reply]
Another solution would be to do what you are doing and pass in the arguments `enter_config_mode=
False` and `exit_config_mode=False` (but if this doesn't work, I would just tell you not to use the method you are currently using as it is intended for configuration operations).

https://github.com/ktbyers/netmiko/blob/develop/netmiko/base_connection.py#L2083

**Reply by andreirapuru** [reply]
How to pass this argument?

**Reply by ktbyers** [reply]
Add it to your `send_config_from_file()` call.

**Reply by andreirapuru** [reply]
Hello @ktbyers.
I used another script with send_command() and it worked for those access-points:

#!/usr/bin/env python
import logging
from netmiko import ConnectHandler
from getpass import getpass
import csv

device_template = {
    "device_type": "cisco_ios",
    "host": "10.10.10.71",
    "username": "dnac",
    "password": getpass(),
}

list_of_devices = csv.DictReader(open('devices_to_configure.csv'))

for row in list_of_devices:
	device_template['device_type'] = 'cisco_ftd'
	device_template['port'] = '22'
	device_template['ip'] = row['ip']
	with ConnectHandler(**device_template) as net_connect:
		output = net_connect.send_command('show version')
	print()
	print(output)
	print()
	log = open('log_file.txt', 'a')
	log.write('\n')
	log.write(device_template['ip'])
	log.write('\n')
	log.write(output)
	log.write('\n')
	net_connect.disconnect()
	print ('====== Logout device', device_template['ip'],' ======')


Thanks a lot for your help in this discussion and for netmiko.
